### PR TITLE
fix golint warnings/errors on package api/types/

### PR DESCRIPTION
**Pull request by WeiZhang555**
Part of #14756
Let me know if anyone has suggestions or questions, thank you!

Signed-off-by: Zhang Wei zhangwei555@huawei.com

### CODE REVIEW

**Comment by WeiZhang555** [comment]
Seems there are some failed test cases, I'll check the code and re-open this PR after fixing all bugs.

**Comment by WeiZhang555** [comment]
I'm confused about the failed test cases since they works fine on my machine. 
For example, Janky report:
`
----------------------------------------------------------------------
FAIL: docker_cli_images_test.go:63: DockerSuite.TestImagesFilterLabel

docker_cli_images_test.go:91:
    c.Fatalf("Expected ids %s,%s got %s", image1ID, image2ID, out)
... Error: Expected ids <no value>,<no value> got 92dd539a6f6d013d175cc764a860e6c791a8f0bb54a8b9c61e4ef25d41fd1649
ac680f88b1f7d83af42a67e581c83fa42d69f07239386528f0298e505c252d33

----------------------------------------------------------------------
PASS: docker_cli_images_test.go:101: DockerSuite.TestImagesFilterSpaceTrimCase	1.552s

----------------------------------------------------------------------
FAIL: docker_cli_images_test.go:21: DockerSuite.TestImagesOrderedByCreationDate

docker_cli_images_test.go:46:
    c.Fatalf("First image must be %s, got %s", id3, imgs[0])
... Error: First image must be <no value>, got ada0863e79cd3d94d845f1eab4d8290876e3d696fe4647c6857de64cdc801918

----------------------------------------------------------------------
PASS: docker_cli_import_test.go:36: DockerSuite.TestImportBadURL	0.102s

PASS: docker_cli_import_test.go:13: DockerSuite.TestImportDisplay	1.722s

PASS: docker_cli_import_test.go:46: DockerSuite.TestImportFile	1.293s
PASS: docker_cli_import_test.go:75: DockerSuite.TestImportFileNonExistentFile	0.038s
PASS: docker_api_info_test.go:10: DockerSuite.TestInfoApi	0.012s

PASS: docker_cli_info_test.go:11: DockerSuite.TestInfoEnsureSucceeds	0.036s

----------------------------------------------------------------------
FAIL: docker_api_inspect_test.go:12: DockerSuite.TestInspectApiContainerResponse

docker_api_inspect_test.go:41:
    c.Fatalf("%s does not exist in response for version %s", key, cs.version)
... Error: Id does not exist in response for version 1.20

----------------------------------------------------------------------
`
but on my machine, the failed test cases mentioned above all passed：
`
PASS: docker_cli_images_test.go:63: DockerSuite.TestImagesFilterLabel	0.982s
PASS: docker_cli_images_test.go:21: DockerSuite.TestImagesOrderedByCreationDate	2.950s
PASS: docker_api_inspect_test.go:12: DockerSuite.TestInspectApiContainerResponse	        0.276s
`

So is there any possibility that it's not my problem? Or could you kindly give me some hints about this?  @icecrime @runcom @jfrazelle

File: api/types/types.go [hunk 1]
```diff
@@ -58,16 +64,18 @@ type ImageHistory struct {
...
 	Untagged string `json:",omitempty"`
 	Deleted  string `json:",omitempty"`
 }
 
+// Image contains response of Remote API:
 // GET "/images/json"
 type Image struct {
 	ID          string `json:"Id"`
-	ParentId    string
+	ParentID    string
```
Comment (vdemeester) [inline api/types/types.go]: As it is *serialized* as JSON, I think it would be better to do something like :

```go
ParentID string `json:ParentId`
```

This way, this limit the possible *breakage* of tools using the API.

File: api/types/types.go [hunk 2]
```diff
@@ -76,14 +84,17 @@ type Image struct {
...
 type GraphDriverData struct {
 	Name string
 	Data map[string]string
 }
 
+// ImageInspect contains response of Remote API:
 // GET "/images/{name:.*}/json"
 type ImageInspect struct {
-	Id              string
+	ID              string
```
Comment (vdemeester) [inline api/types/types.go]: Same as above, as it is *serialized* as JSON, I think it would be better to do something like :

```go
ID string `json:Id`
```

Comment (runcom) [inline api/types/types.go]: :+1:

File: api/types/types.go [hunk 3]
```diff
@@ -140,15 +155,18 @@ type ContainerPathStat struct {
...
 	Processes [][]string
 	Titles    []string
 }
 
+// Version contains response of Remote API:
+// GET "/version"
 type Version struct {
 	Version       string
-	ApiVersion    version.Version
+	APIVersion    version.Version
```
Comment (vdemeester) [inline api/types/types.go]: Same as above, as it is *serialized* as JSON, I think it would be better to do something like :

```go
APIVersion version.Version `json:ApiVersion`
```

File: api/types/types.go [hunk 4]
```diff
@@ -167,11 +186,11 @@ type Info struct {
 	DriverStatus       [][2]string
 	MemoryLimit        bool
 	SwapLimit          bool
-	CpuCfsPeriod       bool
-	CpuCfsQuota        bool
+	CPUCfsPeriod       bool
```
Comment (vdemeester) [inline api/types/types.go]: Same as above, as it is *serialized* as JSON, I think it would be better to do something like :

```go
CPUCfsPeriod bool `json:CpuCfsPeriod`
CPUCfsQuota bool `json:CpuCfsQuota`
```

File: api/types/types.go [hunk 5]
```diff
@@ -189,15 +208,15 @@ type Info struct {
 	NCPU               int
 	MemTotal           int64
 	DockerRootDir      string
-	HttpProxy          string
-	HttpsProxy         string
+	HTTPProxy          string
```
Comment (vdemeester) [inline api/types/types.go]: Same as above, as it is serialized as JSON, I think it would be better to do something like :

```go
HTTPProxy string `json:"HttpProxy"`
HTTPSProxy string `json:"HttpsProxy"`
```

File: api/types/types.go [hunk 6]
```diff
@@ -219,9 +240,10 @@ type ContainerState struct {
 	FinishedAt time.Time
 }
 
+// ContainerJSONBase contains response of Remote API:
 // GET "/containers/{name:.*}/json"
 type ContainerJSONBase struct {
-	Id              string
+	ID              string
```
Comment (vdemeester) [inline api/types/types.go]: Same as above, as it is *serialized* as JSON, I think it would be better to do something like :

```go
ID string `json:"Id"`
```

File: api/types/types.go [hunk 7]
```diff
@@ -258,14 +281,15 @@ type ContainerJSONPre120 struct {
...
+// ContainerConfig is a backcompatibility struct used in ContainerJSONPre120
 type ContainerConfig struct {
 	*runconfig.Config
 
 	// backward compatibility, they now live in HostConfig
 	Memory     int64
 	MemorySwap int64
-	CpuShares  int64
-	Cpuset     string
+	CPUShares  int64
```
Comment (vdemeester) [inline api/types/types.go]: Same as above, as it is *serialized* as JSON, I think it would be better to do something like :

```go
CPUShares bool `json:"CpuShares"`
CPUSet bool `json:"CpuSet"`
```

File: api/types/types.go (hunk 2 above)
Comment (WeiZhang555) [inline api/types/types.go]: @vdemeester Do you mean that only `ID` field should be modified like this, or the `Parent`,'Comment', 'Created' etc fields below should also do these works? 
Same as below :p

Comment (vdemeester) [inline api/types/types.go]: @WeiZhang555 yeah only the one you changed because on ``golint`` needs that (to be sure the serialization keep the same *form*).

File: api/types/types.go [hunk 8]
```diff
@@ -244,13 +266,14 @@ type ContainerJSONBase struct {
...
 
+// ContainerJSON is newly used struct along with MountPoint
 type ContainerJSON struct {
 	*ContainerJSONBase
 	Mounts []MountPoint
 	Config *runconfig.Config
 }
 
-// backcompatibility struct along with ContainerConfig
+// ContainerJSONPre120 is a backcompatibility struct along with ContainerConfig
```
Comment (vdemeester) [inline api/types/types.go]: s/backcompatibility/backward compatibility/ maybe ? :blush:

File: integration-cli/check_test.go [hunk 9]
```diff
@@ -38,8 +38,12 @@ func (s *DockerRegistrySuite) SetUpTest(c *check.C) {
 }
 
 func (s *DockerRegistrySuite) TearDownTest(c *check.C) {
-	s.reg.Close()
-	s.ds.TearDownTest(c)
+	if s.reg != nil {
```
Comment (vdemeester) [inline integration-cli/check_test.go]: Not sure why this change ? It's not really related to the goal of the PR. Did you have issues with integration-cli tests ?

**Comment by vdemeester** [comment]
@WeiZhang555 I think you should rebase to make the integration-cli test passes (I'm pretty sure a fix has been merged :wink: ).

File: api/types/types.go (hunk 8 above)
Comment (WeiZhang555) [inline api/types/types.go]: Yeah, that seems better, will change it :p

File: integration-cli/check_test.go (hunk 9 above)
Comment (WeiZhang555) [inline integration-cli/check_test.go]: Nice catch :+1: . During my development of this PR, this piece of code paniced because one of my bug.
I would remove this to reduce confusion and focus more on the golint warning removement.

**Comment by WeiZhang555** [comment]
@vdemeester Updated, let's see if this commit works better. :smile:

File: integration-cli/check_test.go (hunk 9 above)
Comment (vdemeester) [inline integration-cli/check_test.go]: thx :blush:

**Comment by WeiZhang555** [comment]
Rebased

**Comment by WeiZhang555** [comment]
Rebased again.

**Comment by vdemeester** [comment]
@WeiZhang555 could you take care of my comments about the structure and the ``json:…`` stuff ? This will make those changes less error prone — you shouldn't have to modify the tests that uses the API (like ``inspectField(…)``) because this means it breaks a little bit the API and thus it could break external tools that consume the API (like docker/compose, kubernetes, etc..). :blush:

**Comment by WeiZhang555** [comment]
@vdemeester Yeah, I'm so sorry, I must have missed that, I would add `json` part and have a double check.

**Comment by WeiZhang555** [comment]
@vdemeester Updated, sorry for my careless, I have added the `json` stuff. You are right,  breaking the json API format is bad and we should try hard to avoid this kind of modification unless we have to.
And the code changing with `json` stuff is more reasonable and elegant now, thank you! :smile:

**Comment by vdemeester** [comment]
@WeiZhang555 don't be sorry :stuck_out_tongue_closed_eyes:, this is part a the learning and the contributing process :blush: :wink:.

**Comment by tiborvass** [comment]
@WeiZhang555 really sorry but needs another rebase :(

**Comment by WeiZhang555** [comment]
@tiborvass Rebase done. :)

**Comment by WeiZhang555** [comment]
Rebased

**Comment by jhowardmsft** [comment]
Needs another rebase to pick up the compile error on the Windows daemon build.

**Comment by WeiZhang555** [comment]
All green now. Ping @vdemeester @tiborvass @icecrime @jhowardmsft

**Comment by jhowardmsft** [comment]
LGTM

File: integration-cli/check_test.go (hunk 9 above)
Comment (abronan) [inline integration-cli/check_test.go]: So, should the nil checks be removed or not? :ghost:

File: api/types/types.go [hunk 10]
```diff
@@ -166,11 +185,11 @@ type Info struct {
...
 	MemoryLimit        bool
 	SwapLimit          bool
-	CpuCfsPeriod       bool
-	CpuCfsQuota        bool
+	CPUCfsPeriod       bool `json:"CpuCfsPeriod"`
+	CPUCfsQuota        bool `json:"CpuCfsQuota"`
 	IPv4Forwarding     bool
 	BridgeNfIptables   bool
-	BridgeNfIp6tables  bool
+	BridgeNfIP6tables  bool
```
Comment (vdemeester) [inline api/types/types.go]: This one probably need a ```json:…``` too :sweat_smile:

File: integration-cli/check_test.go (hunk 9 above)
Comment (vdemeester) [inline integration-cli/check_test.go]: I would say no on this particular PR because it should work and worked as is.

File: api/types/types.go (hunk 10 above)
Comment (abronan) [inline api/types/types.go]: Nice catch :+1:

**Comment by WeiZhang555** [comment]
@vdemeester Updated and rebased, thank you! :smile:

**Comment by calavera** [comment]
LGTM. Merging!
